Cubical/Data/Queue/Untruncated2List.agda

```diff
@@ -0,0 +1,164 @@
+{-# OPTIONS --cubical --no-exact-split --safe #-}
+module Cubical.Data.Queue.Untruncated2List where
+
+open import Cubical.Foundations.Everything
+
+open import Cubical.Foundations.SIP
+open import Cubical.Structures.Queue
+
+open import Cubical.Data.Empty as ⊥
+open import Cubical.Data.Unit
+open import Cubical.Data.Sum
+open import Cubical.Data.List
+open import Cubical.Data.Sigma
+open import Cubical.HITs.PropositionalTruncation
+
+open import Cubical.Data.Queue.1List
+
+module _ (A : Type ℓ) (Aset : isSet A) where
+ open Queues-on A Aset
+
+ -- Untruncated 2Lists
+ data Q : Set ℓ where
+   Q⟨_,_⟩ : (xs ys : List A) → Q
+   tilt : ∀ xs ys z → Q⟨ xs ++ [ z ] , ys ⟩ ≡ Q⟨ xs , ys ++ [ z ] ⟩
+
+  -- enq into the first list, deq from the second if possible
+
+ flushEq' : (xs ys : List A) → Q⟨ xs ++ ys , [] ⟩ ≡ Q⟨ xs , rev ys ⟩
+ flushEq' xs [] = cong Q⟨_, [] ⟩ (++-unit-r xs)
+ flushEq' xs (z ∷ ys) j =
+   hcomp
+     (λ i → λ
+       { (j = i0) → Q⟨ ++-assoc xs [ z ] ys i , [] ⟩
+       ; (j = i1) → tilt xs (rev ys) z i
+       })
+     (flushEq' (xs ++ [ z ]) ys j)
+
+ flushEq : (xs ys : List A) → Q⟨ xs ++ rev ys , [] ⟩ ≡ Q⟨ xs , ys ⟩
+ flushEq xs ys = flushEq' xs (rev ys) ∙ cong Q⟨ xs ,_⟩ (rev-rev ys)
+
+ emp : Q
+ emp = Q⟨ [] , [] ⟩
+
+ enq : A → Q → Q
+ enq a Q⟨ xs , ys ⟩ = Q⟨ a ∷ xs , ys ⟩
+ enq a (tilt xs ys z i) = tilt (a ∷ xs) ys z i
+
+ deqFlush : List A → Unit ⊎ (Q × A)
+ deqFlush [] = inl tt
+ deqFlush (x ∷ xs) = inr (Q⟨ [] , xs ⟩ , x)
+
+ deq : Q → Unit ⊎ (Q × A)
+ deq Q⟨ xs , [] ⟩ = deqFlush (rev xs)
+ deq Q⟨ xs , y ∷ ys ⟩ = inr (Q⟨ xs , ys ⟩ , y)
+ deq (tilt xs [] z i) = path i
+   where
+   path : deqFlush (rev (xs ++ [ z ])) ≡ inr (Q⟨ xs , [] ⟩ , z)
+   path =
+     cong deqFlush (rev-snoc xs z)
+     ∙ cong (λ q → inr (q , z)) (sym (flushEq' [] xs))
+ deq (tilt xs (y ∷ ys) z i) = inr (tilt xs ys z i , y)
+
+ Raw : RawQueue
+ Raw = (Q , emp , enq , deq)
+
+ -- We construct an equivalence Q₁≃Q and prove that this is a queue-iso
```

Comment:
Does this imply that the untruncated Queues are a set if A is?

Reply:
Yep. I added the proof.

Reply:
Great!